functional-interfaces-practice/src/FunctionalInterfaces.java

```diff
@@ -0,0 +1,170 @@
+    //NOTE BinaryOperator is better because the return type is the same as the input type.
+
+    //6. Comparator Lambda
+    //Write a Comparator<String> that sorts strings alphabetically ignoring case.
+    Comparator<String> stringComparator = (s1, s2) -> s1.compareToIgnoreCase(s2);
+    //Method Reference
+    Comparator<String> sComparator = String::compareToIgnoreCase;
+
+    //7. Chained Functions
+    //Define two Function<Integer,Integer>: one adds 5, another multiplies by 2.
+    //Use andThen and compose to show different results.
+    Function<Integer, Integer> addFive = x -> x + 5;
+    Function<Integer, Integer> multiplyByTwo = x -> x * 2;
+    //1. add 2. multiply
+    int andThan = addFive.andThen(multiplyByTwo).apply(5);
+    //1. multiply 2. add
+    int compose = addFive.compose(multiplyByTwo).apply(5);
+
+
+    //8. Negating Predicates
+    //Write a Predicate<String> that checks if a string is empty.
+    //Then negate it using .negate().
+    Predicate<String> isEmpty = s -> s.isEmpty();
+    //Method Reference
+    Predicate<String> isEmptyMethodReference = String::isEmpty;
+    Predicate<String> negate = isEmpty.negate();
+
+    //9. Method Reference
+    //Express the following lambda as a method reference:
+    //Function<String,Integer> f = s -> Integer.parseInt(s);
+    Function<String, Integer> f = Integer::parseInt;
+
+    //10. Consumer Chaining
+    //Write two Consumer<String>: one prints the string, another prints its length.
+    //Chain them with andThen.
+    Consumer<String> print = s -> System.out.println(s);
+    //Method Reference
+    Consumer<String> printS = System.out::println;
+    Consumer<String> printLength = s -> System.out.println(s.length());
+
+    Consumer<String> combinedConsumer = print.andThen(printLength);
+
+    //11. BinaryOperator Example
+    //Write a BinaryOperator<Integer> that returns the maximum of two integers.
+    BinaryOperator<Integer> witchIsGreater = (a, b) -> a > b ? a : b;
+
+    //12. Decide the Interface
+    //Given the lambda () -> "Hello", decide whether it’s a Supplier<String> or something else.
+    Supplier<String> printHello = () -> "Hello";
+
+    //13. BiFunction Example
+    //Write a BiFunction<String,String,Integer> that takes two strings and returns the length of their concatenation.
+    BiFunction<String, String, Integer> biFunction = (s1, s2) -> (s1 + s2).length();
+
+    //14. UnaryOperator Example
+    //Write a UnaryOperator<String> that trims a string and converts it to uppercase.
+    UnaryOperator<String> unaryOperator = s -> s.trim().toUpperCase();
+    //trim() --> cut the spaces at the beginning and at the end of the string
+
+    //15. Function Composition Chain
+    //Chain three Function<String,String> lambdas:
+    //add "Hello " prefix
+    //convert to uppercase
+    //add "!" suffix
+    //Apply it to "world".
+    //NOTE
+    // PREFIX ---> adds a string to the START of 's'
+    Function<String, String> addPrefix = s -> "Hello " + s;
```

Comment:
yep, but please note that if someone uses this in a for loop, we are getting into the ugly case of string concatenation within a loop :-)